crates/sage-cloudpath/src/parquet.rs

```diff
@@ -227,6 +233,130 @@ pub fn serialize_features(
     writer.into_inner()
 }
 
+pub fn build_matched_fragment_schema() -> parquet::errors::Result<Type> {
+    let msg = r#"
+        message schema {
+            required byte_array psm_id (utf8);
+            required byte_array fragment_type (utf8);
+            required int32 fragment_ordinals;
+            required int32 fragment_charge;
+            required float fragment_mz_experimental;
+            required float fragment_mz_calculated;
+            required float fragment_intensity;
+        }
+    "#;
+
+    // required float fragment_mz_experimental;
+    // required float fragment_mz_calculated;
+    parquet::schema::parser::parse_message_type(msg)
+}
+
+pub fn serialize_matched_fragments(
+    features: &[Feature]
+) -> Result<Vec<u8>, parquet::errors::ParquetError> {
+
+    let schema = build_matched_fragment_schema()?;
+
+    let options = WriterProperties::builder()
+        .set_compression(parquet::basic::Compression::ZSTD(ZstdLevel::try_new(3)?))
+        .build();
+
+    let buf = Vec::new();
+
+    let mut writer = SerializedFileWriter::new(buf, schema.into(), options.into())?;
+
+    for features in features.chunks(65536) {
+
+        let mut rg = writer.next_row_group()?;
+
+        if let Some(mut col) = rg.next_column()? {
+            let psm_ids= features.iter()
+                .map(|f| vec![f.psm_id.to_string().as_bytes().into();  f.fragments.fragment_ordinals.len()])
+                .flatten()
+                .collect::<Vec<ByteArray>>();
+
+            col.typed::<ByteArrayType>()
+                .write_batch(&psm_ids, None, None)?;
+            col.close()?;
+        }
+
+        if let Some(mut col) = rg.next_column()? {
+            let fragment_types= features.iter()
+                .map(|f| f.fragments.kinds.clone())
+                .flat_map(|each_list| each_list.into_iter().map(|kind| match kind {
+                    Kind::A => { "a".to_string().as_bytes().into()}
```

Comment:
`to_string().as_bytes()` should be elided by directly constructing `b"a"`

Reply:
b"a" is a byte string literal, and the result is a byte array (&[u8; 1]). However, the Parquet library's data type ByteArray accepts a byte slice (&[u8]). In order to resolve this, we can include an additional mapping step as demonstrated below. Currently, I have used "a".as_bytes().into(). Please inform me if any changes need to be made to the code.

```
.map(|kind| match kind {
                    Kind::A => b"a",
                    Kind::B => b"b",
                    Kind::C => b"c",
                    Kind::X => b"x",
                    Kind::Y => b"y",
                    Kind::Z => b"z",
                }).map(|slice| ByteArray::from(&slice[..]))
```